### PR TITLE
Update to spigot 1.17

### PR DESCRIPTION
**Pull request by RainbowDashLabs**
Update to 1.17

### CODE REVIEW

**Review by Aurelien30000 — DISMISSED** [review]

File: LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java [hunk 1]
```diff
@@ -8,14 +8,16 @@
...
+import java.util.List;
 
 public class MobsManager implements IMobManager {
 
-    private static final Collection<IMob> MOBS = new ArrayList<>();
+    private static final List<IMob> MOBS = new ArrayList<>();
     final int currentMineCraftVersion;
 
     public MobsManager() {
+        // v1_17_R1 becomes 1171. Using protocol version should be better, but it is not not supported by Spigot currently.
```
Comment (SirYwell) [inline LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java:20]: WorldEdit exposes the data version used by the platform. We could use that if we want.

File: LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java [hunk 2]
```diff
@@ -35,6 +37,12 @@ public MobsManager() {
         if (currentMineCraftVersion >= 1162) {
             register1_16_2Entities();
         }
+
+        if (currentMineCraftVersion >= 1171) {
+            register1_16_2Entities();
```
Comment (SirYwell) [inline LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java:42]:
```suggestion
            register1_17Entities();
```

File: LandLord-api/src/main/java/biz/princeps/landlord/api/IMobManager.java [hunk 3]
```diff
@@ -24,9 +24,10 @@
...
 
     /**
-     * Gets a collection of all available Mobs.
+     * Gets a list of all available Mobs.
      *
-     * @return a collection of all mobs
+     * @return a list of all mobs
      */
-    Collection<IMob> values();
+    List<IMob> values();
```
Comment (SirYwell) [inline LandLord-api/src/main/java/biz/princeps/landlord/api/IMobManager.java:31]: This is an API/ABI breaking change. Is it required? From what I've seen, it's not necessary to expose that Lists are used internally.

File: LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java [hunk 4]
```diff
@@ -35,6 +37,12 @@ public MobsManager() {
...
+        }
+
+        MOBS.sort(Comparator.comparing(iMob -> iMob.getType().name()));
```
Comment (SirYwell) [inline LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java:45]: Why do we sort them now?

**Review by SirYwell — CHANGES_REQUESTED** [review]

File: LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java (hunk 4 above)
Comment (Aurelien30000) [reply LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java:45]: Sorted entities are better for users imo, they will be sorted alphabetically in guis.

File: LandLord-api/src/main/java/biz/princeps/landlord/api/IMobManager.java (hunk 3 above)
Comment (Aurelien30000) [reply LandLord-api/src/main/java/biz/princeps/landlord/api/IMobManager.java:31]: Gonna revert this.

File: LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java (hunk 2 above)
Comment (Aurelien30000) [reply LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java:42]: Thanks!

File: LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java (hunk 1 above)
Comment (Aurelien30000) [reply LandLord-latest/src/main/java/biz/princeps/landlord/manager/MobsManager.java]: Good idea, I chose to directly use data version, didn't remember about Bukkit unsafe values :D

File: LandLord-latest/src/main/java/biz/princeps/landlord/LandLord.java [hunk 5]
```diff
@@ -91,7 +91,7 @@ protected boolean checkDependencies() {
 
         // Dependency stuff
         final String version = Bukkit.getVersion();
-        if (!version.contains("1.13.2") && !version.contains("1.14") && !version.contains("1.15") && !version.contains("1.16")) {
+        if (!version.contains("1.13.2") && !version.contains("1.14") && !version.contains("1.15") && !version.contains("1.16") && !version.contains("1.17")) {
             haltPlugin("Invalid Spigot version detected! LandLord latest requires 1.13.2/1.14.x/1.15.x/1.16.x, use Legacy version for 1.12.2!");
```
Comment (SirYwell) [inline LandLord-latest/src/main/java/biz/princeps/landlord/LandLord.java:95]: This line doesn't mention 1.17 for now

Comment (Aurelien30000) [reply LandLord-latest/src/main/java/biz/princeps/landlord/LandLord.java]: Oops 😅

**Review by SirYwell — APPROVED** [review]

**Comment by Aurelien30000** [comment]
Currently, the main issue I noticed is caused by an incompatibility between worldguard / spigot and paper shaded classes / dependencies. 

Anyways, there is a workaround in worldguard config:
```java
extra-timings:
    session-handlers: false
```

Another error seems to pop out about eldoutilities:
```java
[06:10:31] [Server thread/ERROR]: Error occurred while enabling Landlord v4.351-SNAPSHOT-022d9b2 (Is it up to date?)
java.lang.ExceptionInInitializerError: null
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:567) ~[?:?]
	at java.lang.Class.getEnumConstantsShared(Class.java:3766) ~[?:?]
	at java.lang.System$2.getEnumConstantsShared(System.java:2170) ~[?:?]
	at java.util.EnumMap.getKeyUniverse(EnumMap.java:754) ~[?:?]
	at java.util.EnumMap.<init>(EnumMap.java:135) ~[?:?]
	at biz.princeps.landlord.eldoutilities.crossversion.functionbuilder.FunctionBuilder.<init>(FunctionBuilder.java:11) ~[?:?]
	at biz.princeps.landlord.eldoutilities.crossversion.functionbuilder.VersionFunctionBuilder.functionBuilder(VersionFunctionBuilder.java:17) ~[?:?]
	at biz.princeps.landlord.eldoutilities.core.EldoUtilities.ignite(EldoUtilities.java:88) ~[?:?]
	at biz.princeps.landlord.ALandLord.onEnable(ALandLord.java:62) ~[?:?]
	at biz.princeps.landlord.LandLord.onEnable(LandLord.java:60) ~[?:?]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:263) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:342) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:480) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at org.bukkit.craftbukkit.v1_17_R1.CraftServer.enablePlugin(CraftServer.java:495) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at org.bukkit.craftbukkit.v1_17_R1.CraftServer.enablePlugins(CraftServer.java:409) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at net.minecraft.server.MinecraftServer.loadWorld(MinecraftServer.java:608) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at net.minecraft.server.dedicated.DedicatedServer.init(DedicatedServer.java:264) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at net.minecraft.server.MinecraftServer.x(MinecraftServer.java:987) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at net.minecraft.server.MinecraftServer.lambda$0(MinecraftServer.java:307) ~[11727.jar:3138-Spigot-66f9d3c-77058c7]
	at java.lang.Thread.run(Thread.java:831) [?:?]
Caused by: java.lang.NullPointerException: Cannot invoke "String.isEmpty()" because the return value of "java.util.regex.Matcher.group(int)" is null
	at biz.princeps.landlord.eldoutilities.crossversion.ServerVersion.extractVersion(ServerVersion.java:93) ~[?:?]
	at biz.princeps.landlord.eldoutilities.crossversion.ServerVersion.getVersion(ServerVersion.java:63) ~[?:?]
	at biz.princeps.landlord.eldoutilities.crossversion.ServerVersion.<clinit>(ServerVersion.java:48) ~[?:?]
	... 23 more
```

**Comment by Aurelien30000** [comment]
@RainbowDashLabs You should figure this out :D

I took a quick look, format apparently doesn't changed.
This is probably caused by a missing patch version such as ``1.17.x``.
```properties
# old
version=1.16.5-R0.1-SNAPSHOT
# new
version=1.17-R0.1-SNAPSHOT
```

**Comment by RainbowDashLabs** [comment]
> 
> 
> @RainbowDashLabs You should figure this out :D
> 
> I took a quick look, format apparently doesn't changed.
> This is probably caused by a missing patch version such as `1.17.x`.
> 
> ```ini
> # old
> version=1.16.5-R0.1-SNAPSHOT
> # new
> version=1.17-R0.1-SNAPSHOT
> ```
> 
> Thanks!

Yeah it kinda way. Checked on a null capture group if it was empty. It's the first minor version update for eldo utils xD Hope we don't find any more suprises in it :D

**Review by Aurelien30000 — DISMISSED** [review]

**Review by Aurelien30000 — APPROVED** [review]

**Pull request merged**